rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+  for two BYOC integrations to have incompatible runtimes.
+
+- **Additive cost assumption**: Collage as per this design assumes the cost of running candidate partitions is additive,
+  plus a small transition penalty. However cache effects can dominate measured latency, particularly for
+  'lightweight' kernels. Thus there may be a **additive error** in the final result:
+
+  > additive_error = measured_latency(collage_partitioning) - sum_{partition} (estimated_latency(partition) + penalty)
+
+  The evolutionary search explored by the paper can help here since it uses measured end-to-end model latency as its
+  cost function, but we're deferring that to future work.
+
+- **Limited search space**: Naively exploring all sub-graphs is O(n!), so we need to constrain the search. The easiest
+  approach is just to limit candidates to sub-graphs of just a few operators. This can mean significantly faster
+  candidates are not explored, yielding a partitioning with high **optimality loss**:
+
+  > optimality_loss = measured_latency(collage_partitioning) - measured_latency(true_optimal_partitioning)
+
+  Though the 'true' optimal partitioning may be infeasible to find, the user may easily discover a high
+  **apparent loss**, eg by comparing the Collage result with a traditional BYOC partitioning result:
+
+  > apparent_loss = measured_latency(collage_partitioning) - measured_latency(users_own_partitioning)
+
+- **Fragile toolchains**: Some BYOC toolchains are intended to be stand-alone compilers in their own right, and have
+  been tuned against common models and include global flags to guide optimizations such as reducing precision. However
+  Collage will only feed these toolchains smaller sub-graphs, thus making the limited search space problem more severe.
+
+- **High variance in lightweight kernels**: Small kernels can have high variance, thus the choice of which toolchain to
+  use can be arbitrary. We probably want to i) validate our variance estimator is accurate, ii) choose a percentile
+  slightly above 50% for the estimated candidate kernel latency, and iii) fall back to hard-coded priorities when the
+  measured variance is too high.
+
+- **Explainability**: It's easy to show the user the final partitioning and estimated times for each kernel, but harder
+  to show why that partitioning won out over all others during search.
+
+- **Does not subsume `partition_for_<toolchain>`**: We don't have any plan to deprecate the existing patterns of each
+  BYOC integration supplying a `partiion_for_<toolchain>` function. If the user has a specific toolchain in mind then
+  making the partition explicit enjoys both faster compilation and can incorporate global optimization passes which
+  Collage cannot currently account for (eg enforcing a particular layout).
+
+## Appendix A: Differences between the paper's implementation and this proposal
```

Comment:
Note to myself : reviewed up to here.

Reply:
Thanks Manupa, I'll hold off working through these until you've finished a first pass.